processor/spanmetricsprocessor/processor.go

```diff
@@ -381,37 +442,72 @@ func (p *processorImp) buildDimensionKVs(serviceName string, span pdata.Span, op
 	return dims
 }
 
-func concatDimensionValue(metricKeyBuilder *strings.Builder, value string, prefixSep bool) {
+func extractResourceAttrsByKeys(serviceName string, keys []Dimension, resourceAttrs pdata.AttributeMap) pdata.AttributeMap {
+	dims := pdata.NewAttributeMap()
+	dims.UpsertString(serviceNameKey, serviceName)
+	for _, ra := range keys {
+		if attr, ok := resourceAttrs.Get(ra.Name); ok {
+			dims.Upsert(ra.Name, attr)
+		} else if ra.Default != nil {
+			// Set the default if configured, otherwise this metric should have no value set for the resource attribute.
+			dims.Upsert(ra.Name, pdata.NewAttributeValueString(*ra.Default))
+		}
+	}
+	return dims
+}
+
+func concatDimensionValue(metricKeyBuilder *strings.Builder, value string) {
 	// It's worth noting that from pprof benchmarks, WriteString is the most expensive operation of this processor.
 	// Specifically, the need to grow the underlying []byte slice to make room for the appended string.
-	if prefixSep {
+	if metricKeyBuilder.Len() != 0 {
 		metricKeyBuilder.WriteString(metricKeySeparator)
 	}
 	metricKeyBuilder.WriteString(value)
 }
 
-// buildKey builds the metric key from the service name and span metadata such as operation, kind, status_code and
+// buildMetricKey builds the metric key from the service name and span metadata such as operation, kind, status_code and
 // will attempt to add any additional dimensions the user has configured that match the span's attributes
 // or resource attributes. If the dimension exists in both, the span's attributes, being the most specific, takes precedence.
 //
 // The metric key is a simple concatenation of dimension values, delimited by a null character.
-func buildKey(serviceName string, span pdata.Span, optionalDims []Dimension, resourceAttrs pdata.AttributeMap) metricKey {
+func (p *processorImp) buildMetricKey(span pdata.Span, resourceAttrs pdata.AttributeMap) metricKey {
 	var metricKeyBuilder strings.Builder
-	concatDimensionValue(&metricKeyBuilder, serviceName, false)
-	concatDimensionValue(&metricKeyBuilder, span.Name(), true)
-	concatDimensionValue(&metricKeyBuilder, span.Kind().String(), true)
-	concatDimensionValue(&metricKeyBuilder, span.Status().Code().String(), true)
+	concatDimensionValue(&metricKeyBuilder, span.Name())
+	concatDimensionValue(&metricKeyBuilder, span.Kind().String())
+	concatDimensionValue(&metricKeyBuilder, span.Status().Code().String())
 
-	for _, d := range optionalDims {
+	for _, d := range p.dimensions {
 		if v, ok := getDimensionValue(d, span.Attributes(), resourceAttrs); ok {
-			concatDimensionValue(&metricKeyBuilder, v.AsString(), true)
+			concatDimensionValue(&metricKeyBuilder, v.AsString())
 		}
 	}
 
 	k := metricKey(metricKeyBuilder.String())
 	return k
 }
 
+// buildResourceAttrKey builds the metric key from the service name and will attempt to add any additional resource attributes
```

Comment:
Are we making the assumption that every resource revolves around service name? If so what the is reasoning behind this

Reply:
OTel defines service.name as a mandatory resource attribute (see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md). Originally, the processor defined service.name as a mandatory attribute and fetched it from the resource attributes then set it as a normal attribute. This didn't make much sense to me so I altered the behaviour so that it's a default resource attribute.